helix-vcs/src/git.rs

```diff
@@ -0,0 +1,197 @@
+    time::{Duration, Instant},
+};
+
+use git2::{Oid, Repository};
+use ropey::Rope;
+use similar::DiffTag;
+
+use crate::{rope::RopeLine, LineDiff, LineDiffs, RepoRoot};
+
+pub struct Git {
+    repo: Repository,
+    /// Absolute path to root of the repo
+    root: RepoRoot,
+    head: Oid,
+
+    /// A cache mapping absolute file paths to file contents
+    /// in the HEAD commit.
+    head_cache: HashMap<PathBuf, Rope>,
+}
+
+impl std::fmt::Debug for Git {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Git").field("root", &self.root).finish()
+    }
+}
+
+impl Git {
+    pub fn head_commit_id(repo: &Repository) -> Option<Oid> {
+        repo.head()
+            .and_then(|gitref| gitref.peel_to_commit())
+            .map(|commit| commit.id())
+            .ok()
+    }
+
+    pub fn discover_from_path(file: &Path) -> Option<Self> {
+        let repo = Repository::discover(file).ok()?;
+        let root = repo.workdir()?.to_path_buf();
+        let head_oid = Self::head_commit_id(&repo)?;
+        Some(Self {
+            repo,
+            root,
+            head: head_oid,
+            head_cache: HashMap::new(),
+        })
+    }
+
+    pub fn root(&self) -> &Path {
+        &self.root
+    }
+
+    fn relative_to_root<'p>(&self, path: &'p Path) -> Option<&'p Path> {
+        path.strip_prefix(&self.root).ok()
+    }
+
+    pub fn read_file_from_head(&mut self, file: &Path) -> Option<&Rope> {
+        let current_head = Self::head_commit_id(&self.repo)?;
+        // TODO: Check cache validity on events like WindowChange
+        // instead of on every keypress ? Will require hooks.
+        if current_head != self.head {
+            self.head_cache.clear();
+            self.head = current_head;
+        }
+
+        if !self.head_cache.contains_key(file) {
+            let relative = self.relative_to_root(file)?;
+            let revision = &format!("HEAD:{}", relative.display());
+            let object = self.repo.revparse_single(revision).ok()?;
+            let blob = object.peel_to_blob().ok()?;
+            let contents = std::str::from_utf8(blob.content()).ok()?;
+            self.head_cache
+                .insert(file.to_path_buf(), Rope::from_str(contents));
+        }
+
+        self.head_cache.get(file)
+    }
+
+    pub fn line_diff_with_head(&mut self, file: &Path, contents: &Rope) -> LineDiffs {
+        let old = match self.read_file_from_head(file) {
+            Some(rope) => RopeLine::from_rope(rope),
+            None => return LineDiffs::new(),
+        };
+        let new = RopeLine::from_rope(contents);
+
+        let mut line_diffs: LineDiffs = HashMap::new();
+        let mut mark_lines = |range: Range<usize>, change: LineDiff| {
+            for line in range {
+                line_diffs.insert(line, change);
+            }
+        };
+
+        let timeout = Duration::from_millis(250);
+        let diff = similar::capture_diff_slices_deadline(
```

Comment:
If I'm reading your code correctly, it looks like you're collecting all the lines into a `Vec` in order to pass them to `capture_diff_slices_deadline()`, so even though this has a deadline set you can still be bottle necked on collecting everything into a `Vec`.

Instead, you could use `capture_diff_deadline()` by wrapping the `Rope` in a type that implements `Index` and returns the lines through indexing.  That way the entire procedure can be deadlined.

I could also make this even easier for you by implementing `Hash` on `RopeSlice` (which is really just an accidental omission on my part in Ropey).

Reply:
We only diff using `similar` initially and now use changesets to incrementally compute the diffs, so it's no longer executed on every keypress. It'll still be nice to optimize it using indexing since we don't want to stall on huge text files on startup.